### PR TITLE
[fmt,folly] backport '/utf-8 only if the compiler is MSVC at build time'

### LINKED ISSUE #40912: [fmt] build failure: clang++: error: no such file or directory: '/utf-8'
### Operating system

Windows 11 23H2 (OS Build 22631.4037)

### Compiler

clang version 18.1.8

### Steps to reproduce the behavior

```Shell
### Versions

Windows 11 23H2 (OS Build 22631.4037)
clang version 18.1.8
cmake version 3.30.3

### Step 1. cmake configure

**Commands**:


git clone --branch 2024.08.23 --single-branch https://github.com/microsoft/vcpkg
.\vcpkg\bootstrap-vcpkg.bat
.\vcpkg\vcpkg install --triplet=x64-windows

cmake -S . -B build --preset use_vcpkg -GNinja -DCMAKE_CXX_COMPILER=clang++ --trace


**Output**:

Pay attention to this argument: `\$<\$<COMPILE_LANGUAGE:CXX>:/utf-8>`.
Looks like it comes from: vcpkg\ports\folly\fix-fmt11-cmake.patch: `string(REPLACE "<$<COMPILE_LANGUAGE:CXX>:/utf-8>" "/utf-8" cflags "${cflags}")`

```shell
C:/dev/vcpkg_error_minimal_example/build/vcpkg_installed/x64-windows/share/fmt/fmt-targets.cmake(60):  set_target_properties(fmt::fmt PROPERTIES INTERFACE_COMPILE_DEFINITIONS FMT_SHARED INTERFACE_COMPILE_FEATURES cxx_std_11 INTERFACE_COMPILE_OPTIONS \$<\$<COMPILE_LANGUAGE:CXX>:/utf-8> INTERFACE_INCLUDE_DIRECTORIES ${_IMPORT_PREFIX}/include )
```

### Step 2. cmake build -> error

**Commands**:

```shell
cmake --build build -- -k 0
```

**Output**:

```shell
[1/2] Building CXX object CMakeFiles/main_cpp.dir/main.cpp.obj
FAILED: CMakeFiles/main_cpp.dir/main.cpp.obj
C:\PROGRA~1\LLVM\bin\CLANG_~1.EXE -DFMT_SHARED -DSPDLOG_COMPILED_LIB -DSPDLOG_FMT_EXTERNAL -DSPDLOG_SHARED_LIB -isystem C:/dev/vcpkg_error_minimal_example/build/vcpkg_installed/x64-windows/include -O0 -g -Xclang -gcodeview -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd /utf-8 -MD -MT CMakeFiles/main_cpp.dir/main.cpp.obj -MF CMakeFiles\main_cpp.dir\main.cpp.obj.d -o CMakeFiles/main_cpp.dir/main.cpp.obj -c C:/dev/vcpkg_error_minimal_example/main.cpp
clang++: error: no such file or directory: '/utf-8'
ninja: build stopped: cannot make progress due to previous errors.
```

### CMakePreset.json

```json
{
    "version": 8,
    "configurePresets": [
        {
            "name": "use_vcpkg",
            "cacheVariables": {
                "CMAKE_TOOLCHAIN_FILE": "vcpkg/scripts/buildsystems/vcpkg.cmake"
            }
        }
    ]
}
```

### vcpkg.json

```json
{
  "dependencies": [
    "spdlog"
  ]
}
```

### CMakeLists.txt

```cmake
cmake_minimum_required(VERSION 3.15)
project(main_cpp)
find_package(spdlog CONFIG REQUIRED)
add_executable(main_cpp main.cpp)
target_link_libraries(main_cpp PRIVATE spdlog::spdlog)
```

### main.cpp

```cpp
int main() {}
```
```


### Failure logs

[1/2] Building CXX object CMakeFiles/main_cpp.dir/main.cpp.obj
FAILED: CMakeFiles/main_cpp.dir/main.cpp.obj
C:\PROGRA~1\LLVM\bin\CLANG_~1.EXE -DFMT_SHARED -DSPDLOG_COMPILED_LIB -DSPDLOG_FMT_EXTERNAL -DSPDLOG_SHARED_LIB -isystem C:/dev/vcpkg_error_minimal_example/build/vcpkg_installed/x64-windows/include -O0 -g -Xclang -gcodeview -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd /utf-8 -MD -MT CMakeFiles/main_cpp.dir/main.cpp.obj -MF CMakeFiles\main_cpp.dir\main.cpp.obj.d -o CMakeFiles/main_cpp.dir/main.cpp.obj -c C:/dev/vcpkg_error_minimal_example/main.cpp
clang++: error: no such file or directory: '/utf-8'
ninja: build stopped: cannot make progress due to previous errors.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by aminya**
Backporting https://github.com/fmtlib/fmt/pull/4159

Fixes #40912

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by miyanyan** [comment]
there's a folly patch when I update fmt to 11.0.2, https://github.com/microsoft/vcpkg/pull/39738/files#diff-33c48021044f60d7e2b48d022de38098102f34806db18005e14549ec0414ab1c
```
diff --git a/CMake/GenPkgConfig.cmake b/CMake/GenPkgConfig.cmake
index 0e93175..1637d9f 100644
--- a/CMake/GenPkgConfig.cmake
+++ b/CMake/GenPkgConfig.cmake
@@ -93,7 +93,7 @@ function(gen_pkgconfig_vars)
   # Set the output variables
   string(REPLACE ";" " " cflags "${cflags}")
   string(REPLACE ";" " " private_libs "${private_libs}")
-
+  string(REPLACE "<$<COMPILE_LANGUAGE:CXX>:/utf-8>" "/utf-8" cflags "${cflags}")
   # Since CMake 3.18 FindThreads may include a generator expression requiring
   # a target, which gets propagated to us through INTERFACE_COMPILE_OPTIONS.
   # Before CMake 3.19 there's no way to solve this in a general way, so we
```
I use `string(REPLACE "<$<COMPILE_LANGUAGE:CXX>:/utf-8>" "/utf-8" cflags "${cflags}")` to avoid cmake error, see https://github.com/facebook/folly/issues/2250, so maybe this patch should be changed also

**Comment by WangWeiLin-MV** [comment]
As @miyanyan said, 

> maybe this patch should be changed also

**Comment by aminya** [comment]
I removed the folly patch. That patch would cause the same fmt error for folly. The current fmt patch is robust under LLVM or MSVC.

**Comment by BillyONeal** [comment]
Has the folly change been submitted to folly?

**Comment by aminya** [comment]
@BillyONeal Created a PR:
https://github.com/facebook/folly/pull/2293

**Comment by WangWeiLin-MV** [comment]
Please resolve the git conflicts.

**Comment by Laro88** [comment]
Comment, this would be nice to have resolved properly, fmt is currently breaking out builds.

**Comment by aminya** [comment]
Conflicts fixed

**Review by WangWeiLin-MV — APPROVED** [review]

**Comment by ysyecust** [comment]
Looking forward to merging it soon. Due to this problem, clang-tidy cannot be used normally on Windows if the fmt library is included.

File: ports/fmt/fix-pass-utf-8-only-if-the-compiler-is-MSVC-at-build.patch
Comment (WangWeiLin-MV) [inline ports/fmt/fix-pass-utf-8-only-if-the-compiler-is-MSVC-at-build.patch:1]: Prefer downloading patches from upstream

```cmake
vcpkg_download_distfile(PATCH_PR_4159
    URLS https://github.com/fmtlib/fmt/commit/f924d16e4789d6bde9def439ac50b778a1536384.patch?full_index=1
    SHA512 5b89b21972db459ea718ea9ad599cc0056856c2e4952baaf901a87a9508b2fe65b13028bf606e42453a11efe81bffc498de4daec7fce05793d8bd4a3b6bed895
    FILENAME fmtlib-fmt-pr-4159-only-pass-utf-8-for-msvc.patch
)
```

File: ports/fmt/portfile.cmake [hunk 1]
```diff
@@ -6,6 +6,7 @@ vcpkg_from_github(
     HEAD_REF master
     PATCHES
         fix-write-batch.patch
+        fix-pass-utf-8-only-if-the-compiler-is-MSVC-at-build.patch # remove in next release
```
Comment (WangWeiLin-MV) [inline ports/fmt/portfile.cmake:9]:
```suggestion
        "${PATCH_PR_4159}"
```

**Review by WangWeiLin-MV — COMMENTED** [review]
Just a suggestion

**Comment by aminya** [comment]
@WangWeiLin-MV I prefer to keep the setup simple. The next version of fmt will remove this patch anyway.

**Review by WangWeiLin-MV — APPROVED** [review]
The port `fmt` and `folly` usage tests pass with the following triplets:

* x64-linux
* x64-windows

**Comment by WangWeiLin-MV** [comment]
Tag reviewed for further review.

**Comment by Karma5s** [comment]
Still getting this error `vcpkg_installed\x64-windows\x64-windows\include\fmt\base.h(458,28): error C2338: static_assert failed: 'Unicode support requires compiling with /utf-8'` when compiling with MSVC following this tutorial
https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-msbuild?pivots=shell-powershell

**Pull request merged**